kv/etcd/mock.go

```diff
@@ -1,71 +1,498 @@
+
+	if err != nil {
+		return nil, err
 	}
 
-	closer := closer.Func(func() error {
-		etcd.Server.Stop()
-		return nil
+	return res.Put(), nil
+}
+
+// Txn implements the Clientv3Facade interface
+func (m *mockKV) Txn(ctx context.Context) clientv3.Txn {
+	return &mockTxn{
+		ctx: ctx,
+		kv:  m,
+	}
+}
+
+// Compact implements the Clientv3Facade interface
+func (m *mockKV) Compact(context.Context, int64, ...clientv3.CompactOption) (*clientv3.CompactResponse, error) {
+	panic("Compact unimplemented")
+}
+
+// Do implements the Clientv3Facade interface
+func (m *mockKV) Do(_ context.Context, op clientv3.Op) (clientv3.OpResponse, error) {
+	m.valuesMtx.Lock()
+	defer m.valuesMtx.Unlock()
+	return m.doInternal(op)
+}
+
+func (m *mockKV) doInternal(op clientv3.Op) (clientv3.OpResponse, error) {
+	if op.IsGet() {
+		return m.doGet(op)
+	} else if op.IsPut() {
+		return m.doPut(op)
+	} else if op.IsDelete() {
+		return m.doDelete(op)
+	} else if op.IsTxn() {
+		return m.doTxn(op)
+	} else {
+		panic(fmt.Sprintf("unsupported operation: %+v", op))
+	}
+}
+
+func (m *mockKV) doGet(op clientv3.Op) (clientv3.OpResponse, error) {
+	kvs := make([]*mvccpb.KeyValue, 0)
+	matching := m.matchingKeys(op, m.values)
+
+	for _, k := range matching {
+		kv := m.values[k]
+		kvs = append(kvs, &kv)
+	}
+
+	res := clientv3.GetResponse{
+		Kvs:   kvs,
+		Count: int64(len(kvs)),
+	}
+
+	return res.OpResponse(), nil
+}
+
+func (m *mockKV) doDelete(op clientv3.Op) (clientv3.OpResponse, error) {
+	matching := m.matchingKeys(op, m.values)
+
+	for _, k := range matching {
+		kv := m.values[k]
+		kv.ModRevision = kv.Version
+
+		m.sendEvent(clientv3.Event{
+			Type: mvccpb.DELETE,
+			Kv:   &kv,
+		})
+
+		delete(m.values, k)
+	}
+
+	res := clientv3.DeleteResponse{Deleted: int64(len(matching))}
+	return res.OpResponse(), nil
+}
+
+func (m *mockKV) doPut(op clientv3.Op) (clientv3.OpResponse, error) {
+	keyBytes := op.KeyBytes()
+	valBytes := op.ValueBytes()
+	key := string(keyBytes)
+
+	var newVal mvccpb.KeyValue
+	oldVal, ok := m.values[key]
+
+	if ok {
+		newVal = oldVal
+		newVal.Version = newVal.Version + 1
+		newVal.ModRevision = newVal.ModRevision + 1
+		newVal.Value = valBytes
+	} else {
+		newVal = mvccpb.KeyValue{
+			Key:            keyBytes,
+			Value:          valBytes,
+			Version:        1,
+			CreateRevision: 1,
+			ModRevision:    1,
+		}
+	}
+
+	m.values[key] = newVal
+	m.sendEvent(clientv3.Event{
+		Type: mvccpb.PUT,
+		Kv:   &newVal,
 	})
 
-	var config Config
-	flagext.DefaultValues(&config)
+	res := clientv3.PutResponse{}
+	return res.OpResponse(), nil
+}
 
-	client := &Client{
-		cfg:    config,
-		codec:  codec,
-		cli:    v3client.New(etcd.Server),
-		logger: logger,
+func (m *mockKV) doTxn(op clientv3.Op) (clientv3.OpResponse, error) {
+	cmps, thens, elses := op.Txn()
+	succeeded := m.evalCmps(cmps)
+
+	var toRun []clientv3.Op
+	if succeeded {
+		toRun = thens
+	} else {
+		toRun = elses
+	}
+
+	responses := make([]*etcdserverpb.ResponseOp, len(toRun))
```

Comment:
```suggestion
	responses := make([]*etcdserverpb.ResponseOp, 0, len(toRun))
```

Otherwise this will allocate the slice with the given length, not only the given capacity. Further down we append to this slice, the appended values would then end up behind `len(toRun)` empty values.